Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -30,167 +30,320 @@
-        _threadRunningLock = PTHREAD_MUTEX_INITIALIZER;
+        return candidateOperation != nil;
     }
-};
 
-extern pthread_key_t g_currentDispatchQueue;
+    void WaitUntilNotSuspended() {
+        std::unique_lock<std::mutex> lock(_lock);
+        while (_isSuspended) {
+            _suspendedCondition.wait(lock);
+        }
+        lock.unlock();
+    }
 
-static inline gpointer CompareExchangePointer(volatile gpointer* dest, gpointer exch, gpointer comp) {
-    return (gpointer)EbrCompareExchange((volatile int*)dest, (long)exch, (long)comp);
-}
+    void ClearCandidateLists() {
+        std::lock_guard<std::mutex> lock(_lock);
+        for (int i = 0; i < NSOperationQueuePriority_Count; i++) {
+            _candidateOperationsToRunNext[i].clear();
+        }
+    }
 
-NSAtomicListRef NSAtomicListSteal(NSAtomicListRef* listPtr) {
-    NSAtomicListRef ret;
-    do {
-        ret = *listPtr;
-    } while (CompareExchangePointer((void**)listPtr, NULL, ret) != ret);
-    return ret;
-}
+    BOOL HasMoreWork() {
+        std::lock_guard<std::mutex> lock(_lock);
 
-void NSAtomicListReverse(NSAtomicListRef* listPtr) {
-    struct NSAtomicListNode* cur = *listPtr;
-    struct NSAtomicListNode* prev = NULL;
-    struct NSAtomicListNode* next = NULL;
+        BOOL toReturn = _HasMoreWorkNoLock();
+
+        if (!toReturn) {
+            // No work left to do. Clear out the thread here. This signals that a new work thread is needed to be created and
+            // current one can end its execution.
+            _thread = nil;
+        }
 
-    if (!cur) {
-        return;
+        return toReturn;
     }
 
-    do {
-        next = cur->next;
-        cur->next = prev;
+    void AddOperation(NSOperation* operation, unsigned int priority, NSOperationQueue* queue, SEL selector) {
+        {
+            std::lock_guard<std::mutex> lock(_lock);
 
-        if (next) {
-            prev = cur;
-            cur = next;
+            StrongId<NSOperation> strongOperation = operation;
+            _operationsHoldingArea[priority].emplace_back(strongOperation);
+            if (_thread == nil || _thread.get().finished) {
+                _thread.attach([[NSThread alloc] initWithTarget:queue selector:selector object:nil]);
+                [_thread start];
+            }
         }
-    } while (next);
 
-    *listPtr = cur;
-}
+        _workAvailableCondition.notify_one();
+    }
 
-NSObject* NSAtomicListPop(NSAtomicListRef* listPtr) {
-    struct NSAtomicListNode* node = *listPtr;
-    if (!node) {
-        return NULL;
+    void SignalAllWorkDone() {
+        _allWorkDoneCondition.notify_one();
     }
 
-    *listPtr = node->next;
+    void WaitUntilAllOperationsAreFinished() {
+        std::unique_lock<std::mutex> lock(_lock);
 
-    NSObject* elt = [[node->elt retain] autorelease];
-    delete node;
-    return elt;
-}
+        while (_HasMoreWorkNoLock()) {
+            _allWorkDoneCondition.wait(lock);
```

Comment:
yeah I thought about it. I'm going to wont fix just because all our logic is currently written in the positive and it seemed more confusing to me to ! it some places. A full rewrite could make use of it better.

---
In reply to: [95450281](https://github.com/Microsoft/WinObjC/pull/1676#discussion_r95450281) [](ancestors = 95450281)